packages/tpetra/core/src/Tpetra_FECrsMatrix_def.hpp

```diff
@@ -138,8 +140,159 @@ void FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::beginFill()  {
+  );
+}
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+LocalOrdinal
+FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::replaceLocalValuesImpl(
+  impl_scalar_type rowVals[],
+  const crs_graph_type& graph,
+  const RowInfo& rowInfo,
+  const LocalOrdinal inds[],
+  const impl_scalar_type newVals[],
+  const LocalOrdinal numElts) const
+{
+  const char tfecfFuncName[] = "FECrsMatrix::replaceLocalValues: ";
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(
+    *fillState_ != FillState::open && *fillState_ != FillState::modify,
+    std::runtime_error,
+    "Cannot replace local values, matrix is not open to fill/modify."
+  );
+  return CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::replaceLocalValuesImpl(
+    rowVals, graph, rowInfo, inds, newVals, numElts
+  );
+}
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+LocalOrdinal
+FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::sumIntoGlobalValuesImpl(
+  impl_scalar_type rowVals[],
+  const crs_graph_type& graph,
+  const RowInfo& rowInfo,
+  const GlobalOrdinal inds[],
+  const impl_scalar_type newVals[],
+  const LocalOrdinal numElts,
+  const bool atomic) const
+{
+  const char tfecfFuncName[] = "FECrsMatrix::sumIntoGlobalValues: ";
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(
+    *fillState_ != FillState::open,
+    std::runtime_error,
+    "Cannot sum in to global values, matrix is not open to fill."
+  );
+  return CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::sumIntoGlobalValuesImpl(
+    rowVals, graph, rowInfo, inds, newVals, numElts, atomic
+  );
+}
 
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+LocalOrdinal
+FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::sumIntoLocalValuesImpl(
+  impl_scalar_type rowVals[],
+  const crs_graph_type& graph,
+  const RowInfo& rowInfo,
+  const LocalOrdinal inds[],
+  const impl_scalar_type newVals[],
+  const LocalOrdinal numElts,
+  const bool atomic) const
+{
+  const char tfecfFuncName[] = "FECrsMatrix::sumIntoLocalValues: ";
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(
+    *fillState_ != FillState::open,
+    std::runtime_error,
+    "Cannot sum in to local values, matrix is not open to fill."
+  );
+  return CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::sumIntoLocalValuesImpl(
+    rowVals, graph, rowInfo, inds, newVals, numElts, atomic
+  );
+}
 
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void
+FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::insertGlobalValuesImpl(
+  crs_graph_type& graph,
+  RowInfo& rowInfo,
+  const GlobalOrdinal gblColInds[],
+  const impl_scalar_type vals[],
+  const size_t numInputEnt)
+{
+  const char tfecfFuncName[] = "FECrsMatrix::insertGlobalValues: ";
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(
+    *fillState_ != FillState::open,
+    std::runtime_error,
+    "Cannot sum in to local values, matrix is not open to fill."
+  );
```

Comment:
I think we need the same checks around putScalar, and FECrsMatrix::setAllToScalar - I found I needed begin/endFill around these functions to get the right result (there were no exceptions thrown without the begin/endFill, just inconsistent serial/parallel results). See https://github.com/trilinos/Trilinos/issues/3070 - looks like there are other functions that need is-open checking.